### ISSUE TITLE
[Bug?]: Breaking change in v4 - `yarn workspace(s)` and `$INIT_CWD`

### ISSUE THREAD

**Issue by kolkinn** [yarnpkg/berry#5750]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

In v2 and v3 using `$INIT_CWD` in global scripts allowed the scripts to enter the workspace that initiated them.

```json
{
  "scripts": {
    "g:build": "cd $INIT_CWD && ...",
    "g:pwd": "echo $INIT_CWD"
  }
}
```

This is still the case if you manually change into the directory of a workspace and call the global script from there.

Doing this using `yarn workspace [name of workspace] g:pwd` however, no longer works the same in v4.
In v2 and v3 it would print the path to the workspace specified in the call to `yarn workspace`.
In v4 it instead prints the path from which the `yarn workspace` call originated (CWD).

I couldn't find it on the list of breaking changes for v4, so I'm not sure if it is intentional.
For me at least the change is unfortunate, since I can no longer reuse global scripts in `yarn workspaces foreach` to do something in each workspace directory. Since the CWD for global scripts is the directory in which the script was defined, there seemingly no longer is a way to get to the directory of the workspace from which the global script was called.

By the way, the link to Sherlock in the GitHub issue template doesn't seem to work at the moment. I had to use archive.org to find a working version of the documentation, and even then the playground didn't work with a 500 Internal Server Error.
I luckilly managed to get it working offline.

### To reproduce

```js repro
await packageJson({
  name: "root",
  scripts: {
      "g:pwd": "pwd && echo $INIT_CWD"
  },
  workspaces: [
    "libs/*"
  ],
});

await packageJson({
  name: "demo"
}, { cwd: "./libs/demo" });

await yarn();

const rootPwd = await yarn("exec", "pwd");
const demoPwd = await yarn("workspace", "demo", "exec", "pwd");

expect(demoPwd).not.toEqual(rootPwd);

const singleWorkspace = await yarn("workspace", "demo", "g:pwd");
expect(singleWorkspace).toEqual(rootPwd + demoPwd);

const multipleWorkspaces = await yarn("workspaces", "foreach", "--all", "run", "g:pwd");
expect(multipleWorkspaces).toContain(rootPwd + rootPwd + rootPwd + demoPwd); // Ignore "Done in Xs Yms" on the last line
```

### Environment

```shell
System:
    OS: macOS 13.5.2
    CPU: (10) arm64 Apple M1 Max
  Binaries:
    Node: 18.16.1 - /private/var/folders/jr/f0flt_ds0rzfc_g50cm8jbfc0000gn/T/xfs-4e01b59a/node
    Yarn: 4.0.0-rc.51 - /private/var/folders/jr/f0flt_ds0rzfc_g50cm8jbfc0000gn/T/xfs-4e01b59a/yarn
    npm: 9.5.1 - ~/.volta/tools/image/node/18.16.1/bin/npm
```

**Comment by psychobolt** [comment]
This seems like a regression. Currently the only workaround is to add the `g:` script to every workspace e.g.

```sh
{
  scripts: {
    "my-script": "g:my-script"
  }
}
```

Then instead of calling `yarn workspace my-workspace g:my-script`, call `yarn workspace my-workspace my-script`...

**Comment by antoncazalet** [comment]
Having the same issue in my project, `$INIT_CWD` is not set to the current workspace in the foreach but the one we are running the command.

**Comment by arcanis** [comment]
Indeed, this change was made a bit more than a year ago (https://github.com/yarnpkg/berry/pull/4817). It's too long for me to remember the full context, but I think the idea was that `INIT_CWD` is supposed to always point to the pwd of the user, so the old behavior was seen as a bug.

I however see the problem ... perhaps we should revert INIT_CWD, deprecate it (in documentation only), and add other variables with more descriptive names 🤔

**Comment by antoncazalet** [comment]
Oh, maybe adding a new environnement variable like `CURRENT_WORKSPACE_PWD` would be helpful in the case.

**Comment by austin-payne** [comment]
Just wanted to add my 2 cents since this thread is active 🙂 

I like the suggestion of a new var `CURRENT_WORKSPACE_PWD` but imo a better name is just `WORKSPACE_CWD`. It would take the place of `INIT_CWD` as always pointing to the workspace of the current script, regardless of whether the script is global or run via `yarn workspaces` etc.

It seems intuitive to me considering we have `PROJECT_CWD`, and the meaning of "project" and "workspace" are clearly defined. `INIT_CWD` is incredibly helpful but seemingly more of a tool to define global scripts, and some additional control over its behavior would be helpful.

As described in this issue, `yarn workspaces ...` in v3 would retain the existing value of INIT_CWD when calling each script. I'm gonna refer to this as "forwarding" the INIT_CWD. If we could choose whether or not INIT_CWD is forwarded, not only for `yarn workspaces ...` but also for `yarn run`, `yarn node`, etc, we'd have a lot of flexibility for creating global scripts.

An example use case:
```
Workspace "bundling" has a global script to centralize various bundling code and dependencies:
"bundle:esbuild": "yarn node ./src/esbuild.ts"

esbuild.ts requires some cli args such as absolute paths to the entry file(s) and output directory

Individual workspaces can reuse this script like so:
"bundle": "yarn esbuild:bundle --entry \"$INIT_CWD/src/index.ts\" --outdir \"$INIT_CWD/dist\""
```

Ideally to simplify the bundle scripts, we should be able to pass relative paths:
```
"bundle": "yarn esbuild:bundle --entry ./src/index.ts --outdir ./dist"

How can we let esbuild.ts know the directory to resolve entry and outdir from?
- in esbuild.ts, `process.env.INIT_CWD` is always the path to the "bundling" workspace
- we need to pass a different env var or implement a cli option like `--root`

"bundle:esbuild": "yarn node ./src/esbuild.ts --root $INIT_CWD"
- esbuild.ts can resolve entry and outdir paths from the value of the `--root` option
```

But imagine we had the ability to tell yarn to forward INIT_CWD:
```
"bundle:esbuild": "yarn node --forward-init ./src/esbuild.ts"
- esbuild.ts does not need a custom env var or cli option
- esbuild.ts can resolve entry and outdir paths from `process.env.INIT_CWD`
```

TLDR:
- new var `WORKSPACE_CWD` would behave as INIT_CWD currently behaves in v4
- the v3 behavior of `yarn workspaces ...` was to forward INIT_CWD, which was clearly helpful
  - with a new option, users could choose whether they want forwarding on a per-script basis
- `yarn run`, `yarn node` etc should also be able to optionally forward INIT_CWD

I'm not sure what the best name would be for this new option, `--forward-init` is just an example. My example use case is just a simplified version of something I'm actually doing in a real project. I usually use the `cd $INIT_CWD && ...` pattern, but it isn't always applicable. If we had an option to control the behavior, then INIT_CWD would be a lot more powerful - we could forward multiple levels deep, have more options than `cd $INIT_CWD && ...`, etc

I could be missing something as well, but curious to see where this leads 😄

**Comment by mikeduminy** [comment]
If I can offer a counter-point. In non-global scripts this is a bit unintuitive when using `yarn workspace x`

```js
await packageJson({
  name: "root",
  workspaces: [
    "packages/*"
  ],
});

await packageJson({
  name: "demo",
}, { cwd: "./packages/demo" });

await yarn()

const demoPwd = await yarn("workspace", "demo", "exec", "pwd")
const demoInitCwd = await yarn("workspace", "demo", "exec", "echo \"$INIT_CWD\"")

// this fails today
expect(demoPwd).not.toEqual(demoInitCwd);
```

I don't particularly mind if this is INIT_CWD or something else, but maybe we could have some indicator of the pwd of the caller. To make this non-breaking maybe we introduce CALLER_CWD or something similar, or fix INIT_CWD and expose WORKSPACE_PWD.
